src/scheduler/FunctionCallClient.cpp

```diff
@@ -1,38 +1,165 @@
+#include <faabric/proto/faabric.pb.h>
 #include <faabric/scheduler/FunctionCallClient.h>
 
 #include <grpcpp/create_channel.h>
 #include <grpcpp/security/credentials.h>
 
 #include <faabric/proto/macros.h>
+#include <faabric/util/queue.h>
+#include <faabric/util/testing.h>
 
 namespace faabric::scheduler {
+
+// -----------------------------------
+// Mocking
+// -----------------------------------
+static std::vector<std::pair<std::string, faabric::Message>> functionCalls;
+
+static std::vector<std::pair<std::string, faabric::Message>> flushCalls;
+
+static std::vector<std::pair<std::string, faabric::BatchExecuteRequest>>
+  batchMessages;
+
+static std::vector<std::pair<std::string, faabric::MPIMessage>> mpiMessages;
+
+static std::vector<std::pair<std::string, faabric::ResourceRequest>>
+  resourceRequests;
+
+static std::unordered_map<std::string,
+                          faabric::util::Queue<faabric::HostResources>>
+  queuedResourceResponses;
+
+static std::vector<std::pair<std::string, faabric::UnregisterRequest>>
+  unregisterRequests;
+
+std::vector<std::pair<std::string, faabric::Message>> getFunctionCalls()
+{
+    return functionCalls;
+}
+
+std::vector<std::pair<std::string, faabric::Message>> getFlushCalls()
+{
+    return flushCalls;
+}
+
+std::vector<std::pair<std::string, faabric::BatchExecuteRequest>>
+getBatchRequests()
+{
+    return batchMessages;
+}
+
+std::vector<std::pair<std::string, faabric::MPIMessage>> getMPIMessages()
+{
+    return mpiMessages;
+}
+
+std::vector<std::pair<std::string, faabric::ResourceRequest>>
+getResourceRequests()
+{
+    return resourceRequests;
+}
+
+std::vector<std::pair<std::string, faabric::UnregisterRequest>>
+getUnregisterRequests()
+{
+    return unregisterRequests;
+}
+
+void queueResourceResponse(const std::string& host, faabric::HostResources& res)
+{
+    queuedResourceResponses[host].enqueue(res);
+}
+
+void clearMockRequests()
+{
+    functionCalls.clear();
+    batchMessages.clear();
+    mpiMessages.clear();
+    resourceRequests.clear();
+    unregisterRequests.clear();
+
+    for (auto& p : queuedResourceResponses) {
+        p.second.reset();
+    }
+    queuedResourceResponses.clear();
+}
+
+// -----------------------------------
+// gRPC client
+// -----------------------------------
 FunctionCallClient::FunctionCallClient(const std::string& hostIn)
   : host(hostIn)
   , channel(grpc::CreateChannel(host + ":" + std::to_string(FUNCTION_CALL_PORT),
                                 grpc::InsecureChannelCredentials()))
   , stub(faabric::FunctionRPCService::NewStub(channel))
 {}
 
-void FunctionCallClient::shareFunctionCall(const faabric::Message& call)
```

Comment:
`shareFunctionCall` no longer exists, instead the batch execute RPC is always used